core/src/main/java/hudson/cli/CLIAction.java

```diff
@@ -91,46 +107,159 @@ public Object getTarget() {
+                        Locale locale = Locale.getDefault();
+                        Charset encoding = Charset.defaultCharset();
+                        final PipedInputStream stdin = new PipedInputStream();
+                        final PipedOutputStream stdinMatch = new PipedOutputStream();
+                        ServerSideImpl(InputStream is, OutputStream os) throws IOException {
+                            super(is, os);
+                            stdinMatch.connect(stdin);
+                        }
+                        @Override
+                        protected void onArg(String text) {
+                            args.add(text);
+                        }
+                        @Override
+                        protected void onLocale(String text) {
+                            for (Locale _locale : Locale.getAvailableLocales()) {
+                                if (_locale.toString().equals(text)) {
+                                    locale = _locale;
+                                    return;
+                                }
+                            }
+                            LOGGER.log(Level.WARNING, "unknown client locale {0}", text);
+                        }
                         @Override
-                        protected void main(Channel channel) throws IOException, InterruptedException {
-                            // capture the identity given by the transport, since this can be useful for SecurityRealm.createCliAuthenticator()
-                            channel.setProperty(CLICommand.TRANSPORT_AUTHENTICATION, Jenkins.getAuthentication());
-                            channel.setProperty(CliEntryPoint.class.getName(),new CliManagerImpl(channel));
+                        protected void onEncoding(String text) {
+                            try {
+                                encoding = Charset.forName(text);
+                            } catch (UnsupportedCharsetException x) {
+                                LOGGER.log(Level.WARNING, "unknown client charset {0}", text);
+                            }
+                        }
+                        @Override
+                        protected void onStart() {
+                            ready();
+                        }
+                        @Override
+                        protected void onStdin(byte[] chunk) throws IOException {
+                            stdinMatch.write(chunk);
+                        }
+                        @Override
+                        protected void onEndStdin() throws IOException {
+                            stdinMatch.close();
+                        }
+                        @Override
+                        protected void handleClose() {
+                            ready();
+                            Thread t = runningThread.get();
+                            if (t != null) {
+                                t.interrupt();
+                            }
```

Comment:
Should it also ensure the `sddinMatch` is closed? Just to avoid leaking something